### PR TITLE
[11.0][MIG] sale_exception

### PR DESCRIPTION
**Pull request by simahawk**

### CODE REVIEW

**Comment by simahawk** [comment]
missing dependency on https://github.com/OCA/server-tools/pull/1025.

@pedrobaeza I lost some OCA Transbot commits because of some bad conflicts on rebase cleanup. Can we rely on transbot to update them again when merged or should I try to recover all of them?

**Comment by pedrobaeza** [comment]
As this is a new branch, it will be a new project, so you have to put the correct translations. The best option is: copy translation files from 10.0, make a commit, and squash it with latest "OCA Transbot commit".

As a side note, these conflicts happened when dirty merge/pull operations have been done in the branch in the past. Fortunately, we can fix them here for the future.

**Comment by simahawk** [comment]
ok then, I'll drop all of them and get back from v10. Thanks

**Comment by simahawk** [comment]
@pedrobaeza translations recovered (no transbot commit to squash with tho). 
About build: it complains about base_exception not found but I added this https://github.com/simahawk/sale-workflow/blob/a5cec451f89af26433a650d7a8bd5cd83c913856/oca_dependencies.txt
Is that correct?

**Comment by pedrobaeza** [comment]
The problem is not about base_exception, but another. Please check Travis logs.

**Comment by simahawk** [comment]
@pedrobaeza ah sorry, I looked at an old build that was still failing for dependency issue :(

**Review by StefanRijnhart — APPROVED** [review]
Tests are green, translations are salvaged! Perhaps squash your own commits @simahawk?

File: oca_dependencies.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# base_exception v11 https://github.com/OCA/server-tools/pull/1025
+server-tools https://github.com/akretion/server-tools 11.0-mig-base_exception
```
Comment (LoisRForgeFlow) [inline oca_dependencies.txt]: Please do not forget to remove this.

**Review by LoisRForgeFlow — CHANGES_REQUESTED** [review]

**Comment by StefanRijnhart** [comment]
Travis no like :-(
```
2017-12-05 16:25:17,964 5514 ERROR openerp_test odoo.modules.registry: Model base.exception.test.purchase.line has no table.
2017-12-05 16:25:17,964 5514 ERROR openerp_test odoo.modules.registry: Model base.exception.test.purchase has no table.
```

**Comment by simahawk** [comment]
I know. I'll get back on this later. Anyway the error is related to base_exception that has been merged into server-tools w/out build issues (as far as I see).

Ahh, I guess that it's because the test models are registered only w/ `--test-enable` or `ODOO_TEST_ENABLE` env var. But, modules from other repos are installed "normally".
So, it works on server-tools branch but not here  :cry:

**Comment by simahawk** [comment]
It's really strange because on runbot is not failing and the base module is installed as well w/out test enabled:
```
/home/odoo/OCB-11.0/odoo-bin -d openerp_template --log-level=info --stop-after-init --init base_exception,sale,base
```
On travis I see:
```
 /home/travis/OCB-11.0/odoo-bin -d openerp_test --db-filter=^openerp_test$ --stop-after-init --log-level info --log-handler openerp.tools.yaml_import:DEBUG --test-enable --init sale_exception
```
Maybe is due to the fact that those models are inside `tests` folder :thinking:

**Comment by mourad-ehm** [comment]
I had a same problem with purchase exception https://github.com/OCA/purchase-workflow/pull/467

**Comment by simahawk** [comment]
@mourad-ehm I think is because of this https://github.com/OCA/server-tools/pull/1025#discussion_r155192855
That's one of the reason why I usually put them into `models/`.

**Comment by JordiBForgeFlow** [comment]
@simahawk PR https://github.com/OCA/server-tools/pull/1111 fixes the problem in base_exception

**Comment by simahawk** [comment]
@jbeficent test fixed, thank ;)
@lreficent oca deps updated.
@pedrobaeza ready for another review and merge? :wink:

File: sale_exception/__init__.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# -*- coding: utf-8 -*-
```
Comment (ghost) [inline sale_exception/__init__.py]: can be removed

File: sale_exception/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
```
Comment (ghost) [inline sale_exception/__manifest__.py]: can be removed too

File: sale_exception/models/__init__.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# -*- coding: utf-8 -*-
```
Comment (ghost) [inline sale_exception/models/__init__.py]: can be removed

File: sale_exception/models/sale.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
```
Comment (ghost) [inline sale_exception/models/sale.py]: same

File: sale_exception/models/sale.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+# Copyright 2011 Akretion, Camptocamp, Sodexis
+# Copyright 2017 Akretion, Camptocamp
```
Comment (ghost) [inline sale_exception/models/sale.py]: 2018

File: sale_exception/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+ 'author': "Akretion, Sodexis, Camptocamp, Odoo Community Association (OCA)",
+ 'website': 'http://www.akretion.com',
+ 'depends': ['sale', 'base_exception'],
+ 'license': 'AGPL-3',
+ 'data': [
+     'data/sale_exception_data.xml',
+     'wizard/sale_exception_confirm_view.xml',
+     'views/sale_view.xml',
+ ],
+ }
```
Comment (ghost) [inline sale_exception/__manifest__.py]: I think 4 space indent increase readability

File: sale_exception/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+# Copyright 2011 Akretion, Camptocamp, Sodexis
+# Copyright 2017 Akretion, Camptocamp
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{'name': 'Sale Exception',
+ 'summary': 'Custom exceptions on sale order',
+ 'version': '11.0.1.0.0',
+ 'category': 'Generic Modules/Sale',
+ 'author': "Akretion, Sodexis, Camptocamp, Odoo Community Association (OCA)",
+ 'website': 'http://www.akretion.com',
```
Comment (ghost) [inline sale_exception/__manifest__.py]: specify a link to `https://github.com/OCA/sale-workflow`

File: sale_exception/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
+# Copyright 2011 Akretion, Camptocamp, Sodexis
+# Copyright 2017 Akretion, Camptocamp
```
Comment (ghost) [inline sale_exception/__manifest__.py]: 2018

File: sale_exception/models/sale.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    _inherit = 'exception.rule'
+
+    rule_group = fields.Selection(
+        selection_add=[('sale', 'Sale')],
+    )
+    model = fields.Selection(
+        selection_add=[
+            ('sale.order', 'Sale order'),
+            ('sale.order.line', 'Sale order line'),
+        ])
```
Comment (ghost) [inline sale_exception/models/sale.py]: move `)` parenthesis to new line

File: sale_exception/tests/__init__.py [hunk 11]
```diff
@@ -0,0 +1,5 @@
+# -*- coding: utf-8 -*-
```
Comment (ghost) [inline sale_exception/tests/__init__.py]: can be removed

File: sale_exception/tests/test_sale_exception.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
+# -*- coding: utf-8 -*-
```
Comment (ghost) [inline sale_exception/tests/test_sale_exception.py]: same

File: sale_exception/tests/test_sale_exception.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+# Copyright 2011 Akretion, Camptocamp, Sodexis
+# Copyright 2017 Akretion, Camptocamp
```
Comment (ghost) [inline sale_exception/tests/test_sale_exception.py]: 2018

File: sale_exception/wizard/__init__.py [hunk 14]
```diff
@@ -0,0 +1,2 @@
+# -*- coding: utf-8 -*-
```
Comment (ghost) [inline sale_exception/wizard/__init__.py]: can be removed

File: sale_exception/wizard/sale_exception_confirm.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
```
Comment (ghost) [inline sale_exception/wizard/sale_exception_confirm.py]: same

File: sale_exception/wizard/sale_exception_confirm.py [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+# Copyright 2011 Akretion, Camptocamp, Sodexis
+# Copyright 2017 Akretion, Camptocamp
```
Comment (ghost) [inline sale_exception/wizard/sale_exception_confirm.py]: 2018

**Review by ghost — COMMENTED** [review]
LGTM
micro remarks

**Comment by simahawk** [comment]
@oleksandrpaziuk tnx! all comments addressed.

**Review by ghost — APPROVED** [review]
Great work!

**Review by StefanRijnhart — APPROVED** [review]

File: sale_exception/wizard/sale_exception_confirm.py [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+
+class SaleExceptionConfirm(models.TransientModel):
+    _name = 'sale.exception.confirm'
+    _inherit = ['exception.rule.confirm']
+
+    related_model_id = fields.Many2one('sale.order', 'Sale')
+
+    @api.multi
+    def action_confirm(self):
+        self.ensure_one()
```
Comment (leemannd) [inline sale_exception/wizard/sale_exception_confirm.py:15]: Question, `self.ensure_one()` in multi ? Do we want to be able to call the wizard on multiple SO ?

**Review by leemannd — COMMENTED** [review]
I have only one question before approval

File: sale_exception/wizard/sale_exception_confirm.py (hunk 17 above)
Comment (simahawk) [reply sale_exception/wizard/sale_exception_confirm.py:15]: It's a button on a wizard: you won't use it on multiple SO and in any case you can relate only one SO per wiz, so... :)

Comment (StefanRijnhart) [reply sale_exception/wizard/sale_exception_confirm.py:15]: It's upstream API https://github.com/OCA/server-tools/blob/11.0/base_exception/wizard/base_exception_confirm.py#L38

**Comment by simahawk** [comment]
@StefanRijnhart tnx for reviewing: can you hit the button? :stuck_out_tongue_winking_eye:

**Comment by StefanRijnhart** [comment]
I hit that big and powerful button, and it worked.

**Pull request merged**